src/main/java/com/salesforce/storm/spout/dynamic/metrics/DropwizardRecorder.java

```diff
@@ -0,0 +1,147 @@
+
+        this.keyBuilder = new KeyBuilder(getMetricRecorderTaskIdPrefix());
+
+    }
+
+    private String getMetricRecorderTaskIdPrefix() {
+        // We allow this option so that if you're using Graphite you can keep your key total down
+        if (!spoutConfig.containsKey(SpoutConfig.METRICS_RECORDER_ENABLE_TASK_ID_PREFIX)) {
+            return null;
+        }
+
+        // We have the key, assuming it's a Boolean we'll return it
+        final Object taskIdCfgValue = spoutConfig.get(SpoutConfig.METRICS_RECORDER_ENABLE_TASK_ID_PREFIX);
+        if (taskIdCfgValue instanceof Boolean && (Boolean) taskIdCfgValue) {
+            return "task" + topologyContext.getThisTaskIndex();
+        }
+
+        return null;
+    }
+
+    @Override
+    public void countBy(MetricDefinition metric, long incrementBy, Object... metricParameters) {
+        if (!counters.containsKey(metric)) {
+            counters.put(metric, topologyContext.registerCounter(keyBuilder.build(metric, metricParameters)));
+        }
+
+        counters.get(metric).inc(incrementBy);
+    }
+
+    /**
+     * Assign an arbitrary value.
+     *
+     * This implementation uses a {@link Histogram} because we felt it was closest to the arbitrary value assignment previously done with
+     * the {@link MultiAssignableMetric}.
+     *
+     * This particularly method should be considered experimental at this point as we may revise it before finalizing the release.
+     *
+     * @param metric metric definition.
+     * @param value value to be assigned.
+     * @param metricParameters when a {@link MetricDefinition} supports interpolation on it's key, for example "foo.{}.bar" the {}
+     */
+    @Override
+    public void assignValue(MetricDefinition metric, Object value, Object... metricParameters) {
+        if (!(value instanceof Number)) {
+            throw new IllegalArgumentException("Supplied value must be an instance of Number.");
+        }
+
+        if (!histograms.containsKey(metric)) {
+            histograms.put(metric, topologyContext.registerHistogram(keyBuilder.build(metric, metricParameters)));
+        }
+
+        histograms.get(metric).update(
+            ((Number) value).longValue()
+        );
+    }
+
+    @Override
+    public void startTimer(MetricDefinition metric, Object... metricParameters) {
+        timerManager.start(keyBuilder.build(metric, metricParameters));
+    }
+
+    @Override
+    public long stopTimer(MetricDefinition metric, Object... metricParameters) {
+        final long elapseMs = timerManager.stop(keyBuilder.build(metric, metricParameters));
+        recordTimer(metric, elapseMs, metricParameters);
+        return elapseMs;
+    }
+
+    @Override
+    public void recordTimer(MetricDefinition metric, long elapsedTimeMs, Object... metricParameters) {
+        if (!timers.containsKey(metric)) {
+            timers.put(metric, getTopologyContext().registerTimer(keyBuilder.build(metric, metricParameters)));
```

Comment:
I'm not super familiar w/ the new metrics system in storm.  Under the old system you had to register all of your metrics upfront during the open() phase.  Not being able to support dynamic metrics was generally the reason why we built this abstraction in the first place.

So with that said... is it now possible to register them adhoc after the topology is up and running?

Reply:
So if it IS possible to register metrics dynamically, if my memory serves me, at least for the old metric system, there was a way to ask the topologyContext for a metric instance by key.  If that's still the case, it may mean that keeping instances of the metrics locally to this class in the hashmap may no longer be required.

Reply:
ooh so here's another possible issue.  With the old metric recorder, we didn't ensure thread safety into these methods, under the assumption that if two threads tried to create the same metric at roughly the same time, we might lose a recorded stat, but the end result not much of a harm.

Now if we can register metrics adhoc, if two threads attempt to register the same metric, how does topologyContext handle that?  Under the old metric system trying to register the same metric key twice resulted in an exception.

Reply:
Sounds good, looks like you have a good grasp on how this works.  Yeah I'd swap the map implementations out, otherwise looks good.